### PR TITLE
Refactor errorFromResponse

### LINKED ISSUE #247: Make errorFromResponse parameters optional
I was uprading an API from Apollo Server 2 to Apollo Server 4 when I noticed that I needed to add extra parameters in the `errorFromResponse` method:

https://github.com/apollographql/datasource-rest/blob/fc74bdbc9415bc233eca5037222987e4d2ec033a/src/RESTDataSource.ts#L355-L363

I am using this method in an overriden implementation of `fetch`. I first tried to only use `response` and `parsedbody` but the typescript compiler was complaining about missing parameters.

I went through the implementation and I noticed that we are not using those. So I just used the `as any` assertion trick to bypass the typescript issue, but I don't like this solution, so I decided to contribute in the base code in order to turn `request` and `url` into optional parameters.

PR : https://github.com/apollographql/datasource-rest/pull/246

### PR DESCRIPTION
**Pull request by lotmek**
## Description
- Turn `url` and `request` into optional parameters
- Refactor the way to add `code` in the `extensions` object

Fixes #247

### CODE REVIEW

**Comment by trevor-scheer** [comment]
I think this is a good change, especially since the concrete method here doesn't make use of those params. Thinking out loud because at first glance, this seems like a breaking change. Explaining why it _isn't_.

The signature of the method does not impose any requirements on the signature in concrete class definitions that extend `RESTDataSource` and override this method. That means an overridden `errorFromResponse` method can have _any_ signature, meaning users who currently extend `errorFromResponse` will not be affected by this change. This isn't the first time I've "discovered" this, but usually surprises me.

Additionally, users currently calling this method are unaffected - they won't see any change for already including parameters which become optional (though the other direction optional -> required would be problematic).

Related thought worth capturing that doesn't impact this PR:
`throwIfResponseIsError` and `errorFromResponse` are "linked" by their types. If you override `errorFromResponse` and provide a different signature (specifically, an extended one), you're going to be surprised when `throwIfResponseIsError` calls your method. I'm actually not sure how to fix this.

```ts
class MyDataSource extends RESTDataSource {
  override async errorFromResponse({
    response,
    parsedBody,
  }: {
    response: FetcherResponse;
    parsedBody: string;
    extra: string; // not a TS error, but this won't exist if I don't override `throwIfResponseIsError`
  }): Promise<GraphQLError> {
    return new GraphQLError('...');
  }
}
```

File: src/RESTDataSource.ts [hunk 1]
```diff
@@ -356,18 +356,20 @@ export abstract class RESTDataSource {
...
+    const code = codeByStatus.get(response.status);
+
     return new GraphQLError(`${response.status}: ${response.statusText}`, {
       extensions: {
-        ...(response.status === 401
-          ? { code: 'UNAUTHENTICATED' }
-          : response.status === 403
-          ? { code: 'FORBIDDEN' }
-          : {}),
+        ...(code && { code }),
```
Comment (trevor-scheer) [inline src/RESTDataSource.ts:372]: This change looks correct to me but without a test validating your change I'd rather just leave it as is. I don't think the ternaries here are so complex that it necessitates cleanup.

My preference is to revert this, but if you insist then I'd require a test.

**Review by trevor-scheer — CHANGES_REQUESTED** [review]
Please add a `patch` changeset describing your changes. You can run the changeset tool via `npx changeset` and follow the prompts.

File: src/RESTDataSource.ts (hunk 1 above)
Comment (lotmek) [reply src/RESTDataSource.ts:372]: Hi @trevor-scheer the codes are already tested in those two cases:
- [throws a FORBIDDEN error when the response status is 403](https://github.com/apollographql/datasource-rest/blob/f6106f5e417fb85656643d9dcd15505e3195368e/src/__tests__/RESTDataSource.test.ts#L1596-L1618)
- [throws an UNAUTHENTICATED error when the response status is 401](https://github.com/apollographql/datasource-rest/blob/f6106f5e417fb85656643d9dcd15505e3195368e/src/__tests__/RESTDataSource.test.ts#L1572C11-L1594)

I am just adding an extra assertion in this third test case ([throws an ApolloError when the response status is 500](https://github.com/apollographql/datasource-rest/blob/f6106f5e417fb85656643d9dcd15505e3195368e/src/__tests__/RESTDataSource.test.ts#L1620-L1641)), just to make sure the property is not being added:
https://github.com/apollographql/datasource-rest/blob/410213bb2caadb2860ef2bbfa918e0e7011d7d82/src/__tests__/RESTDataSource.test.ts#L1641

**Review by trevor-scheer — APPROVED** [review]

**Pull request merged**